### DISCUSSION TITLE
Add ability to force rebuild when using dockerfile

### DISCUSSION THREAD

**Discussion by koriwi** [louislam/dockge#267] [Feature Request]
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

API

### 🔖 Feature description

I sometimes build little customized images. But when i use `build:` instead of `image:` in the compose.yaml i always have to stop&down, delete the image and start again in Dockge to get the updated version.

### ✔️ Solution

The update button could rebuild (`--build` equivalent) for services which have a `build:` instead of an `image:`.

### ❓ Alternatives

Add an extra build/rebuild button

**Comment by Yann-J** [comment]
Fully support this!

The UI currently lets you run `docker-compose pull` for a selected stack, but if you are customizing your images, you would instead need `docker-compose build --pull` to trigger a rebuild of your images, including a pull of base images...

**Comment by leouzz** [comment]
This!

**Comment by jgardner-qha** [comment]
I came here looking for this. Not that it's difficult to do it manually, but having a build & pull button would be fantastic!

**Reply by kylmp** [reply]
How do you do it manually?

**Reply by kylmp** [reply]
Nevermind, you can go to the terminal (button in top right) and navigate to where the Dockerfile is then run `docker compose build --pull`

Yes it would be great to have a button to do this in the UI.

**Comment by femabeh** [comment]
and a way to change the docker command thats used when started a container

**Comment by hermannx5** [comment]
I would also love this feature

**Comment by Doc-Steve** [comment]
Count me in!

**Comment by doctore74** [comment]
Absolutely searching for this feature!!!

Also the possibility to edit the Dockerfile in edit mode would be useful.

**Comment by tannisroot** [comment]
Yeah, perhaps Dockerfile could be implemented in a way similar to the way .env is now, as a separate text field, maybe accesible only behind a toggle. And if it's enabled, you get the rebuild button.

**Comment by maulvi** [comment]
bumps

**Comment by knumsi** [comment]
This would be perfect for my homelab

**Comment by AlphaCraft9658** [comment]
I have at least one service that doesn't provide its own Docker images, and I always have to build it on my own. This would be very convenient to have, when I need to rebuild it for a newer release. Maybe, just maybe, it could also detect git repositories and allow for pulling them in the UI for updates. It would also be great if it could detect git repositories in subfolder or would allow them to be declared or something in that manner. That would be super convenient!

**Comment by apeman76** [comment]
bump

**Comment by Qtax** [comment]
This would be great

**Comment by Angrite** [comment]
Bump

**Comment by TacoV** [comment]
Thinking about the implementation. Some configurations with multiple services would use both Pull and Build buttons. Scanning the compose files for which buttons to show feels too comlpicated.
Would it be too much clutter to _always_ have the button, but in the same dropdown as Stop & Inactive?

**Reply by TacoV** [reply]
Will work on a PR but waiting for answer - maybe it's wiser to wait some days/weeks
https://github.com/cmcooper1980/dockge/discussions/20#discussioncomment-15257274

**Comment by ghost** [comment]
Bump

**Comment by realaxl** [comment]
I fully support this request. 
Took me a while to find there's accually no automation build in the Web UI for quick rebuilds for stacks with Dockerfile.

**Comment by TacoV** [comment]
Just to crowdsource this, what would the best practice action be the button performs? `docker compose up -d --build`? Or, as the OP says, down -> delete -> up?

**Comment by Lucidic15** [comment]
Bumping this!

**Comment by klaboor** [comment]
What would "Build" button do if dockerfile does not instruct so? I don't think additional button is necessary. Just honor compose structure during deploy. Scan is the way. Build image if dockerfile instructs so upon deploy, otherwise just deploy. Same for Update button.

**Comment by WunderWungiel** [comment]
Bump!

**Comment by Aerya** [comment]
Hi, added to my Enhanced fork.

**Reply by ukos-git** [reply]
What kind of fork is it — a small patch/feature addition, or a broader set of changes (enhanced fork)? And do you intend to upstream it back via PR, or is it staying independent?

**Reply by Aerya** [reply]
Enhanced rework of Dockge with many features as set [in the readme](https://github.com/Aerya/Dockge-Enhanced#what-sets-dockge-enhanced-apart): restic backup, Trivy scans, updates (with 24h rollback), etc.
I wont PR, I prefer it stays independent. You might give it a try, it's up t odate with the main branch of Dockge (automated workflow).

**Reply by Lucidic15** [reply]
I switched to Portainer and have been much happier.

**Reply by Aerya** [reply]
Used to use it too, with Dockhand, Komodo, Arcane, and got back to Dockge. This was as easy as I needed, except for some features I added :)

**Reply by hermannx5** [reply]
Looks great, thereis just one feature I am missing (or at least not seeing): the update check like in Hamphh's fork. I see that one can update, but I do not see whether an update is available...

**Reply by Aerya** [reply]
Discord, Apprise notifications and tags on stacks names.
